Unencrypted keys in user directory for single-password access (from #1399)

Timvrakas opened the pull request:
NOTE: Continuation of #1399

OK here it is!

Adds new encryption system that uses unencrypted keys the the encrypted user partition, to allow for easier encrypted login. `edit-chroot -u` or `mount-chroot -u` will create the keys, and they will be detected from then on.

I think edit-chroot and mount-chroot are good to go. I tested them on my C720
I am unsure if (or how) I should edit main.sh and add an option to set this up on creation of a chroot, for now you have to run edit-chroot -u or mount-chroot -u to create the keys.

I'm still new to all this (crouton, git, bash, life) so feel free to offer comments, questions, concerns, criticisms, critiques, complaints.

Thanks!

dnschneid commented:
Thanks for working on this!

I don't think there needs to be a new parameter, though.  When you go to encrypt the chroot, and only if you specify the `-k` parameter, you should be allowed to leave the passphrase empty.  crouton should confirm that you want to leave the passphrase in the clear, and note that this should only be done if the keyfile is located in encrypted storage (such as your user profile) or on removable media.  When you go to mount the chroot, the chroot should automatically find the unwrapped key (as it already does for -k).

Timvrakas commented:
@dnschneid I will get on that. It shouldn't take so long now that i'm familiar with the code.

On an off topic git-related tangent, If I want to roll back my current commit, and start fresh (I think that would be easiest), should I `git revert` or `git reset --hard` and then `push --force`? I looked it up, but I'm unsure which is correct. 

Thanks

dnschneid commented:
If you want to start fresh and throw away the work, `git reset --hard origin/master` followed by `git push origin encryption --force-with-lease` will safely update github.

Timvrakas commented:
OK thanks for the tip.

The question I have with the -k parameter is that if .ecryptfs is a pointer for the moved keyfile, then the only location for the keys is in the user directory, which is lost if the user account is deleted. I would prefer to have the keys in 2 places, both encrypted in the folder of the chroot and decrypted in the user home (this is especially true with separate-partition).

Any suggestions for this?

dnschneid commented:
That's kind of the point of moving the key around.  You can manually back up your key somewhere else (and then point to it if need-be by passing -k to enter-chroot), but for security, crouton shouldn't be silently duplicating your keys in multiple places.

tonyxue commented on host-bin/mount-chroot:
```diff
@@ -145,10 +150,28 @@ for NAME in "$@"; do
-        # Detect the key file
+        # Detect unencrypted keys
+        if [ -z "$UKEYFILE" ]; then
+            UKEYFILE="/home/chronos/user/crouton/$NAME.chkey"
```
Why make ```~/crouton``` visible? If it is for storing keys, I think ```~/.crouton``` will be more suitale. Just like ```~/.ssh```

Timvrakas committed:
New style of crouton passwordless encryption

mount-chroot now accepts blank passwords (after prompt), and will
save padded unencrypted keys instead of wrapped keys. requires -k flag to specify custom location for keys.

Timvrakas commented:
Is that the sort of change you were looking for? I guess that makes more sense.
FYI the previous commit was moved to timvrakas/old-encryption
WARNING: The above not tested, I will not have access to my chrome book until later tomorrow.

dnschneid commented on host-bin/mount-chroot:
```diff
@@ -64,11 +66,6 @@ promptNewPassphrase() {
     [ -t 0 ] && stty -echo
     while [ -z "$passphrase" ]; do
         read -r passphrase
-        if [ -z "$passphrase" ]; then
```
Actually, I think this should still be enforced if keyfile wasn't set by the user.

dnschneid commented on host-bin/mount-chroot:
```diff
@@ -77,6 +74,19 @@ promptNewPassphrase() {
         fi
+        if [ -z "$passphrase" ]; then
+            echo 'You did not specify a passphrase! Your keys will not be encrypted.
+            You should only do this if you have specified to store the keys in your
```
```
      echo -n \
'You did not specify a passphrase.  Your keys will not be encrypted.
Only do this if you are confident that your keys are stored in a secure location.
Do you want to continue? [y/N] ' 1>&2
```

dnschneid commented on host-bin/mount-chroot:
```diff
@@ -77,6 +74,19 @@ promptNewPassphrase() {
+            Enter "Yes" to confirm, or "No" to set a passphrase' 1>&2
+            read -r response
+            echo 1>&2
+            if [ "$response" = "N*" -o "$response" = "n*" ]; then
```
Default to no: `if [ "${response#[Yy]" = "$response" ]; then`

dnschneid commented on host-bin/mount-chroot:
```diff
@@ -185,29 +195,30 @@ for NAME in "$@"; do
+                echo | cat - "$wrappedkey" "$wrappedfnek" > "$KEYFILE"
+            else
+                echo > "$KEYFILE"
+                echo -n '$$$$$$$$$$$$$$$$' >> "$KEYFILE"
```
Dollar signs in shell scripts scare me.  How about `.`'s instead?

dnschneid replied:
Also, can we combine this into one call to echo, so it's atomic?

dnschneid commented on host-bin/mount-chroot:
```diff
@@ -185,29 +195,30 @@ for NAME in "$@"; do
+                echo -n '$$$$$$$$$$$$$$$$' >> "$KEYFILE"
+                echo -n "$key" >> "$KEYFILE"
+                echo -n "$fnek" >> "$KEYFILE"
+                echo -n '$$$$$$$$$$$$$$$$' >> "$KEYFILE"
```
The first 16 characters of a wrapped key are the hex signature, so it'd be good to consistently put the padding at the beginning of each key.

I do like this approach overall.

dnschneid commented on host-bin/mount-chroot:
```diff
@@ -216,32 +227,81 @@ $passphrase" | ecryptfs-wrap-passphrase "$wrappedfnek" -
             tail -c 160 "$KEYFILE" | head -c 80 > "$wrappedkey"
             tail -c 80 "$KEYFILE" > "$wrappedfnek"
 
+            PAD1="`cat "$wrappedkey" | head -c 16`"
```
echo, not cat.  But you can also do:

`if [ "${wrappedkey#................}" != "$PAD1" -a "${wrappedfnek#................}" != "$wrappedfnek" ]; then`

dnschneid replied:
Oops, nevermind, you're totally right.

But you CAN pass the file directly to head.

dnschneid commented on host-bin/mount-chroot:
```diff
@@ -216,32 +227,81 @@ $passphrase" | ecryptfs-wrap-passphrase "$wrappedfnek" -
+                    echo -n "$fnek" >> "$KEYFILE"
+                    echo -n '$$$$$$$$$$$$$$$$' >> "$KEYFILE"
+                elif [ -n "$NOENC" -a -n "$NOPASS" ]; then
+                    echo "Changing from nothing to nothing? You must be high-security..."
```
Haha.  Alternatively: `echo "Applying double-ROT13 encryption to keys..."`

dnschneid commented:
But yes, this was pretty much how I was expecting it to behave (kudos for the very clean way of handling the unwrapped keys).

Timvrakas commented:
I still have a problem. They 63-char keys are coming back to haunt me. They are not properly handled. But It should be good otherwise.

Timvrakas commented on host-bin/mount-chroot:
```diff
@@ -60,22 +62,32 @@ addtrap "stty echo 2>/dev/null"
-    while [ -z "$passphrase" ]; do
+    while [ -z "$passphrase" -a -z "$NOPASS" ]; do
         read -r passphrase
-        if [ -z "$passphrase" ]; then
```
About this... I wasn't sure if the warning about secure location was enough, or the blank passphrase should be linked to the -k flag. Right now they are independent features. I get the secure location thing, but seeing as the crouton default is no encryption at all, it feels like overkill.

dnschneid replied:
The point of the `-e` flag is to make your chroot install more secure.  If we allow blank passwords without further configuration, then people will assume that even with a blank passphrase, things are fine (they will ignore the later warnings).  Unless you see actual value in allowing co-located unencrypted passphrases, I'd rather disallow it to avoid any chance of a false sense of security.

dnschneid commented on host-bin/mount-chroot:
```diff
@@ -185,29 +199,27 @@ for NAME in "$@"; do
+                echo | cat - "$wrappedkey" "$wrappedfnek" > "$KEYFILE"
+            else
+                echo -n "
+................$key................$fnek" >> "$KEYFILE"
```
Truncate, not append

dnschneid commented on host-bin/mount-chroot:
```diff
@@ -60,22 +62,34 @@ addtrap "stty echo 2>/dev/null"
+            echo 1>&2
+            if [ "${response#[Yy]}" = "$response" ]; then
+                continue
+			else
```
indentation

Timvrakas replied:
Sorry, that was gedit trying to use tabs instead of spaces.
Now that I mention it, should I be using something more advanced than gedit to do this type of work?

dnschneid replied:
Not really.  I use vim, but you'd do just as well configuring gedit to not insert tabs.

dnschneid commented on host-bin/mount-chroot:
```diff
@@ -216,32 +228,78 @@ $passphrase" | ecryptfs-wrap-passphrase "$wrappedfnek" -
             tail -c 80 "$KEYFILE" > "$wrappedfnek"
 
+            PAD1="`head -c 16 "$wrappedkey"`"
+            PAD2="`tail -c 16 "$wrappedfnek"`"
```
this now needs to be `head`...I would look at the rest of the key handling (especially below) to make sure it's up-to-date with the new ...xxx...yyy format.

Timvrakas replied:
Hey in my defense I caught this before you pointed it out :)

dnschneid replied:
Serves me right for checking things as you're pushing new commits :)

dnschneid commented on host-bin/mount-chroot:
```diff
@@ -216,32 +228,78 @@ $passphrase" | ecryptfs-wrap-passphrase "$wrappedfnek" -
+            PAD1="`head -c 16 "$wrappedkey"`"
+            PAD2="`tail -c 16 "$wrappedfnek"`"
+
+            if [ "$PAD1" = '................' ] && [ "$PAD2" = '................']; then
```
-a is fine here

dnschneid commented on host-bin/mount-chroot:
```diff
@@ -185,29 +199,27 @@ for NAME in "$@"; do
+                echo | cat - "$wrappedkey" "$wrappedfnek" > "$KEYFILE"
+            else
+                echo -n "
+................$key................$fnek" > "$KEYFILE"
```
Can you put the 16 periods into a variable at the top?  Counting them all each time is error-prone.

dnschneid commented on host-bin/mount-chroot:
```diff
@@ -216,32 +228,78 @@ $passphrase" | ecryptfs-wrap-passphrase "$wrappedfnek" -
+                    echo -n "
+................$key................$fnek" > "$KEYFILE"
+                elif [ -n "$NOENC" -a -n "$NOPASS" ]; then
+                    echo "Changing from nothing to nothing? You must be high-security..."
```
I guess the point I was making with the alternate wording was that I'd rather not direct snark at users (each one of them is a special flower and whatever).

Timvrakas replied:
Oh okay, ROT-13 it is :)

dnschneid commented:
Is hexdump still giving you a bad byte count, even with the `32/1` update?  It seemed fixed to me.

Timvrakas commented:
The issue with the hexdump is that old keys (made before the fix) are still too short. I will have to do line-by-line separation.

dnschneid commented:
Good point.  I don't think you need to do line-by-line though.  Something like:

If first 16 characters of key is ....., it's unencrypted (don't check fnek to confirm).  Then you can easily extract the keys (which are just hex):

```
contents="`tail -n+1 keyfile`"
key="${keyfile#${DOTS}}"
key="${key%%.*}"
fnek="${keyfile##*.}"
```

Timvrakas commented:
OK I finished testing and worked out a few bugs. It all works for the most parts.
TODO:
1. Deal with the malformed (short) keys (as mentioned above)
2. The -ee change password should check for request password.
3. Link empty-password to -k flag (as mentioned above)

Timvrakas commented:
Well this is going to sound lame, but I just lost the completed above TODO. Somebody pressed the spacebar..... grr. I will have it done again in a few days.

dnschneid commented:
Whoops!  No rush :)

Timvrakas commented:
Haha, well it looks like I got side-tracked for a long time....
Anyway now I have to deal with the new echo_tty functions

dnschneid commented on host-bin/mount-chroot:
```diff
@@ -60,27 +64,46 @@ addtrap "stty echo 2>/dev/null"
-        fi
         echo_tty ''
-        echo_tty -n 'Please confirm your passphrase: '
+        echo -n 'Please confirm your passphrase: ' 1>&2
```
This should be echo_tty

dnschneid commented on host-bin/mount-chroot:
```diff
@@ -60,27 +64,46 @@ addtrap "stty echo 2>/dev/null"
+            else
+                echo_tty \
+'You tried to specify no passphrase, but to do that,
+you must store you keyfile in a custom location with -k.'
```
```
Passphrases cannot be empty unless the keyfile
is stored separately from the chroot using -k.
```

dnschneid commented on host-bin/mount-chroot:
```diff
@@ -60,27 +64,46 @@ addtrap "stty echo 2>/dev/null"
-    while [ -z "$passphrase" ]; do
+    NOPASS=''
+    if [ ! "$CHROOTSRC/.ecryptfs" = "$KEYFILE" ]; then
+        ENABLENOPASS='y'
```
Does ENABLENOPASS need to be a global variable if you redetermine it each time?

dnschneid commented on host-bin/mount-chroot:
```diff
@@ -60,27 +64,46 @@ addtrap "stty echo 2>/dev/null"
-    [ -t 0 ] && stty -echo
-    while [ -z "$passphrase" ]; do
+    NOPASS=''
+    if [ ! "$CHROOTSRC/.ecryptfs" = "$KEYFILE" ]; then
```
!=

dnschneid commented on host-bin/mount-chroot:
```diff
@@ -60,27 +64,46 @@ addtrap "stty echo 2>/dev/null"
+                if [ "${response#[Yy]}" = "$response" ]; then
+                    continue
+                else
+                NOPASS='y'
```
indentation

dnschneid commented on host-bin/mount-chroot:
```diff
@@ -186,63 +209,112 @@ for NAME in "$@"; do
+                    echo | cat - "$wrappedkey" "$wrappedfnek" > "$KEYFILE"
+                    unset key fnek
+                elif [ -z "$NOENC" -a -n "$NOPASS" ]; then
+                    #Unwrap previously wrapped passphrase
```
space after # (sorry for nits :) )

dnschneid commented:
This is shaping up very well!

dnschneid commented:
The last thing I'd like to see is additions to the 17-encryption tests to handle no passwords (both with -k and without -k specified).

Timvrakas committed:
Further changes to passwordless encryption (as requested by @dnschneid).

dnschneid commented on host-bin/mount-chroot:
```diff
@@ -60,27 +63,46 @@ addtrap "stty echo 2>/dev/null"
-    echo_tty -n "Choose an encryption passphrase for $NAME: "
-    [ -t 0 ] && stty -echo
-    while [ -z "$passphrase" ]; do
+    NOPASS=''
```
now it's a local variable, so add `local`

dnschneid commented:
Considering how critical this is to users' data, I can't merge this unless there are tests in place to prove it works.  Could you add tests to [17-encryption](https://github.com/Timvrakas/crouton/blob/encryption/test/tests/17-encryption) to stress the new code paths?

Timvrakas committed:
Changed NOPASS to local.

Timvrakas commented:
OK, I'm still figuring out the testing architecture, But I will figure it out eventually.

dnschneid commented:
@Timvrakas shouldn't be too complex (you can run the encryption tests on your device by dropping the repo outside your chroot and running `./test/run.sh 17`).  If you need help, swing by \#crouton-devel for advice.  It'd be awesome to get this feature merged, but we have to be careful.

gbrld commented:
Much simpler to just store some long random password in the home folder, then modify mount-chroot to read it out.  Additionally I recommend moving all the scripts from /usr/local/bin to within the eCryptfs home folder (i.e. /home/chronos/user/crouton/) and making a symlink to the chroot storage (i.e. /home/chronos/user/chroot) 

```
mount-chroot:
...
PWFILE="/home/chronos/user/crouton/pwfile"
...
            if [ -e "$PWFILE" ]; then
                passphrase="`cat $PWFILE`"
            else
                echo_tty -n "Enter encryption passphrase for $NAME: "
                [ -t 0 ] && stty -echo
                if [ -z "$passphrase" ]; then
                    read -r passphrase
                fi
                [ -t 0 ] && stty echo
                echo_tty ''
            fi
...
```

dnschneid commented:
> Much simpler to just store some long random password in the home folder, then modify mount-chroot to read it out.

The password would have to have as many bits of entropy as the key itself, assuming the wrapping function actually can use that much data.  Seems like one way or another this would weaken the chroot protection somewhat, compared to storing the key directly in encrypted storage.  Of course, it's still better (and more convenient) than the current approach.

> Additionally I recommend moving all the scripts from /usr/local/bin to within the eCryptfs home folder (i.e. /home/chronos/user/crouton/) and making a symlink to the chroot storage (i.e. /home/chronos/user/chroot)

This is also very prudent.  To be clear, you have to run the scripts directly from the encrypted storage.
